enhancements/console/dynamic-plugins.md

```diff
@@ -0,0 +1,371 @@
+enable or disable parts of the UI based on features in the cluster and is
+typically driven by the resources found during API discovery. Dynamic plugins
+will also be able to declare console feature flags and will only load if the
+flag is enabled. For instance, the KubeVirt plugin will only load if the
+`VirtualMachine` resource is present.
+
+### Error Handling
+
+Console will guard against runtime errors in plugins. All plugin components
+will be surrounded by [React error boundaries](https://reactjs.org/docs/error-boundaries.html).
+This prevents an uncaught error from causing the application to white-screen
+and break.
+
+Additionally, the console can contain a `?disable-plugins` query parameter.
+When this parameter is present, no dynamic plugins are loaded.
+
+### Risks and Mitigations
+
+**Risk**: Patternfly updates with breaking API changes can break plugins.
+
+**Mitigation**: Plugins can declare a semver range for compatibility. The core
+console team will coordinate breaking changes with the plugin teams so that a
+new operator version will be available with fixes when the corresponding
+OpenShift version GAs. We will not ship any breaking Patternfly changes in a
+z-stream.
+
+**Risk**: Plugins are not sandboxed and can make any API request on behalf of the user.
+
+**Mitigation**: Plugins are not enabled by default. A cluster admin must opt-in.
+
+## Design Details
+
+### Test Plan
+
+In some ways, dynamic plugins will simplify how we test since each plugin has
+its own repo. Plugins like KubeVirt and OCS have special requirements that can
+be managed in their own merge queues.
+
+The core console repo will run e2e tests on a sample plugin that exercises the
+dynamic plugin extension points. The tests will install the sample plugin and
+verify that the contributions to different areas of the UI work as expected.
+We'll make the sample plugin available as a reference for other teams
+developing dynamic plugins.
+
+Manual testing of specific plugins like OCS and KubeVirt will continue as
+before, including regression testing for any plugin migrated from static to
+dynamic.
+
+### Graduation Criteria
+
+Initially, this will be an internal API. We will reevaluate as we receive
+implementation feedback from plugin teams and as the dynamic plugin model
+matures.
+
+Static plugins are already a supported feature. Any existing static plugin that
+is migrated to a dynamic plugin will need to have the same support level.
+
+### Upgrade / Downgrade Strategy
+
+We'll need to make sure any static plugin we remove has a dynamic plugin ready.
+OLM index images where operators are locked to a specific OpenShift version
+will allow us to do this. We can make sure the right operator level is required
+for the release where we remove the static plugin.
+
+A plugin might be unavailable for a brief window during a cluster upgrade if
+the console version and operator are at different versions.
```

Comment:
i think this answers my question above

Reply:
Why is this window brief?  Is there something that coordinates core OpenShift and external OLM operator updates?  I know the OLM folks are also working on a version/dependency declaration framework, so it's possible that it's not possible to make strong statements around this sort of thing today, but that we will be able to make stronger statements in the future.